### PR TITLE
feature: Explore Data page redesign

### PR DESCRIPTION
**Pull request by bhongy**

### CODE REVIEW

File: .node-version [hunk 1]
```diff
@@ -1 +1 @@
-16.10.0
+18
```
Comment (bhongy) [inline .node-version:1]: sneak in this change, so our node version manager switches node version automatically

File: apps/web/tailwind.config.js [hunk 2]
```diff
@@ -95,21 +95,17 @@ module.exports = {
       },
       transparent: 'transparent',
     },
-    fontFamily: {
-      montserrat: ['Montserrat', ...defaultTheme.fontFamily.sans],
-      sourceCode: ['Source Code Pro', ...defaultTheme.fontFamily.mono],
-    },
```
Comment (bhongy) [inline apps/web/tailwind.config.js:101]: We have the same thing in `theme.extend` config so this is not necessary.

File: apps/web/tailwind.config.js [hunk 3]
```diff
@@ -72,14 +69,17 @@ module.exports = {
...
         'toast-swipe-out': 'toast-swipe-out 100ms ease-out forwards',
       },
+      backgroundColor: {
+        'venice-red-btn': 'salmon',
+        'venice-green-btn': VeniceTheme.green.darkened,
+      },
+      boxShadow: {
+        'venice-green-glow': `0px 0px 8px 4px ${VeniceTheme.greenGlow}`,
+        'venice-black-drop-shadow': `0px 2px 4px 0px ${VeniceTheme.dropShadow}`,
+      },
```
Comment (bhongy) [inline apps/web/tailwind.config.js:79]: Move to extend so we can use tailwind defaults too

File: apps/web/tailwind.config.js [hunk 4]
```diff
@@ -14,16 +14,13 @@ module.exports = {
     // '../../integrations/**/*.tsx',
     // '../../packages/engine-frontend/**/*.tsx',
 
+    '../../packages/ui/**/*.tsx',
```
Comment (bhongy) [inline apps/web/tailwind.config.js:17]: So tailwind generate utility classes for files in `packages/ui`

File: apps/web/pages/export.tsx [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+      disabled={isCopied}
+      onClick={copyToClipboard}
+      className="shrink-0 gap-1 disabled:opacity-100">
+      <Icon className="h-4 w-4 fill-current" />
+      <span>{text}</span>
+    </Button>
+  )
+}
+
+function useFakeData(query?: string): {
```
Comment (bhongy) [inline apps/web/pages/export.tsx:224]: Remove this once we hook up to the real data.

File: apps/web/pages/export.tsx [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+}
+
+function CopyTextButton(props: CopyTextButtonProps) {
+  const [isCopied, setCopied] = useState(true)
+  const {icon: Icon, text} = isCopied
+    ? {icon: CheckCircleFilledIcon, text: 'Copied'}
+    : {icon: CopyTextIcon, text: 'Copy'}
+
+  async function copyToClipboard() {
+    console.log('copy!')
```
Comment (bhongy) [inline apps/web/pages/export.tsx:197]: @bhongy remove this

File: .node-version (hunk 1 above)
Comment (tonyxiao) [reply .node-version:1]: ah thanks for catching that, I missed it.

File: apps/web/pages/export.tsx [hunk 7]
```diff
@@ -0,0 +1,339 @@
...
+
+const STUB_API_KEY = 'key_01GQHZ4F7GCSFKFWPRZ833PGXP'
+
+function formatCsvQueryUrl(
+  table: string,
+  option: {download: boolean} = {download: false},
+) {
+  const url = new URL('/api/sql', window.location.origin)
+  const params = new URLSearchParams({
+    apiKey: STUB_API_KEY,
```
Comment (bhongy) [inline apps/web/pages/export.tsx:325]: @tonyxiao could you help with getting the correct `apiKey` here?

File: apps/web/pages/export.tsx [hunk 8]
```diff
@@ -0,0 +1,339 @@
...
+import {useEffect, useMemo, useState} from 'react'
+import {ExternalLink} from '../components/ExternalLink'
+import {PageHeader} from '../components/PageHeader'
+import {PageLayout} from '../components/PageLayout'
+import {browserSupabase} from '../contexts/common-contexts'
+
+const PREVIEW_LIMIT = 8
+const DOWNLOAD_LIMIT = 100
+
+export default function Page() {
```
Comment (bhongy) [inline apps/web/pages/export.tsx:31]: @tonyxiao I didn't add `getServerSideProps`—might look into it as a follow-up. Admittedly I'm still not super familiar with next's SSR.

File: apps/cli/README.md [hunk 9]
```diff
@@ -0,0 +1,5 @@
+# Admin Query
```
Comment (bhongy) [inline apps/cli/README.md:1]: @tonyxiao this is unrelated to this PR?

**Comment by tonyxiao** [comment]
Ah right, should push to staging. Good catch

File: packages/ui/components/Progress.tsx [hunk 10]
```diff
@@ -0,0 +1,32 @@
+import {twMerge} from 'tailwind-merge'
+
+interface CircularProgressProps {
+  // pass fill-{color} to change the indicator color
+  // pass text-{color} to change the track color
+  className?: string
+}
+
+export function CircularProgress(props: CircularProgressProps) {
```
Comment (bhongy) [inline packages/ui/components/Progress.tsx:9]: Follow naming (`CircularProgress`) from https://mui.com/material-ui/react-progress/

File: apps/web/pages/export.tsx [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+
+import type {GetServerSideProps} from 'next'
+import Link from 'next/link'
+import {useEffect, useState} from 'react'
+import {PreviewResult, usePreviewQuery} from '../components/export'
+import {ExternalLink} from '../components/ExternalLink'
+import {PageHeader} from '../components/PageHeader'
+import {PageLayout} from '../components/PageLayout'
+
+const PREVIEW_LIMIT = 8
```
Comment (iwasrobbed) [inline apps/web/pages/export.tsx:27]: We likely want a label that says e.g. `8 of 200 shown`, or `8 results shown out of 200 total results total` since this could be confusing as to why only a small number are showing

I'll mock it out

File: apps/web/pages/export.tsx [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+}
+
+export const getServerSideProps: GetServerSideProps<ServerSideProps> = async (
+  ctx,
+) => {
+  const {serverGetUser} = await import('../server')
+  const user = await serverGetUser(ctx)
+
+  const {z} = await import('@usevenice/util')
+  const apiKey = z.string().parse(user?.user_metadata?.['apiKey'])
```
Comment (iwasrobbed) [inline apps/web/pages/export.tsx:41]: If `apiKey` is optional, we should have a fallback scenario for when it gets passed into the UI (and log as an error into Sentry). At some point we should have a way to regenerate it in account settings too

Alternatively a question for @tonyxiao, should it be optional? Also feels a little weird that even `user` is optional

File: packages/ui/components/Select.tsx [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+// TODO use SelectScrollUpButton and SelectScrollDownButton
+export const SelectContent = forwardRef<ContentRef, ContentProps>(
+  function SelectContent(props, ref) {
+    const {children, className, ...contentProps} = props
+    return (
+      <Radix.SelectPortal>
+        <Radix.SelectContent
+          ref={ref}
+          className={twMerge(
+            'min-w-[8rem] animate-slide-down overflow-hidden rounded-lg border border-venice-black-300 bg-venice-black-500 px-1 py-2 text-sm drop-shadow-md animate-in fade-in-80',
```
Comment (iwasrobbed) [inline packages/ui/components/Select.tsx:46]: The only UX nit I have here is it might not be apparent that there are items needing to scroll into view unless you're familiar with the schema. Could we give this a larger default height or handle this in another way?

e.g. with `Resource` selected, tapping the select list only shows
Screen Shot 2023-02-07 at 8 24 33 AM

when really there are more items to display if the user scrolls up
Screen Shot 2023-02-07 at 8 25 41 AM

File: apps/web/pages/export.tsx [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+              </SelectContent>
+            </Select>
+            {/* TODO fade in & out the loader */}
+            {preview.isFetching && <CircularProgress className="h-5 w-5" />}
+          </div>
+          <div className="flex gap-2">
+            <Button
+              variant="primary"
+              className="gap-1"
+              disabled={!selectedTable}
```
Comment (iwasrobbed) [inline apps/web/pages/export.tsx:78]: Nit: may also want to disable it if `results.length` is `0`

**Review by iwasrobbed — COMMENTED** [review]
Looks great overall! Really like the way the preview table & select list look and the page feels & looks nice

🐒  I tried to chaos monkey the UX from a beginner's mind perspective and left some comments for areas that came up

File: apps/web/pages/export.tsx [hunk 15]
```diff
@@ -0,0 +1,253 @@
...
+  return (
+    <PageLayout title="Explore Data">
+      <PageHeader title={['Explore Data', 'CSV']} />
+      <div className="p-6">
+        <div className="grid grid-cols-[1fr_auto]">
+          <div className="flex items-center gap-4">
+            <Select value={selectedTable} onValueChange={selectTable}>
+              <SelectTrigger
+                className="max-w-[10rem]"
+                placeholder="Select a table…"
```
[the remark this answers is not in the document]
Comment (iwasrobbed) [reply apps/web/pages/export.tsx:61]: @tonyxiao fyi this is the issue @bhongy ran into when trying to do the above pre-population:

> Ideally, I think we can preselect a table and renders it server-side but I ran into an error `window.location.origin` doesn't exist during SSR (we need to format the CSV download URL). I punt on that and will circle back to it.
cc/ Tony if you have some idea and can help with the SSR vs window.location.origin issue.

File: packages/ui/components/Select.tsx (hunk 13 above)
Comment (bhongy) [reply packages/ui/components/Select.tsx:46]: It's not a nit at all! It's confusing. That's the UX improvement I mentioned (during stand-up) for `Select` element that I plan to do.

The pop-up has collision detection and limits the height when it hits the top of the pop-up boundary (the document in this case). We will show an indicator (arrow) that there're more items.

File: apps/web/pages/export.tsx (hunk 11 above)
Comment (bhongy) [reply apps/web/pages/export.tsx:27]: 💯 implemented 37c433c

File: apps/web/pages/export.tsx (hunk 14 above)
Comment (bhongy) [reply apps/web/pages/export.tsx]: Good catch!

@iwasrobbed updated in 4f2a3d5 which makes sense for the button. However, for the inline link, it still looks like a link. I tried opacity treatment and it didn't feel right for an inline link. I tried changing the color to white when it's disabled (to make it like a text) which feels **really** wrong when it the data becomes available. I think I'll settle to what we have. wdyt?

File: apps/web/components/export/PreviewResult.tsx [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+        <h3 className="text-venice-gray-muted">{title}</h3>
+        {action}
+      </div>
+    </Card>
+  )
+}
+
+function LoadingPreviewResult() {
+  return (
+    <ul className="w-full animate-pulse will-change-auto">
```
Comment (bhongy) [inline apps/web/components/export/PreviewResult.tsx:111]: @iwasrobbed I'm not sure I'm in love with the pulse animation. Let me know if prefer to have this without animation.

File: apps/web/pages/export.tsx (hunk 14 above)
Comment (iwasrobbed) [reply apps/web/pages/export.tsx]: What you have works for me where it's just a disabled text link!

File: apps/web/components/export/PreviewResult.tsx (hunk 16 above)
Comment (iwasrobbed) [reply apps/web/components/export/PreviewResult.tsx:111]: Think it looks like a user would expect for now, so let's :shipit: I rarely, rarely see this loading indicator anyways so it's already an edge case design

File: apps/web/pages/export.tsx [hunk 17]
```diff
@@ -0,0 +1,253 @@
...
+      disabled={isCopied}
+      onClick={copyToClipboard}
+      className="shrink-0 gap-1 disabled:opacity-100">
+      <Icon className="h-4 w-4 fill-current" />
+      <span>{text}</span>
+    </Button>
+  )
+}
+
+function formatCsvQueryUrl({
```
[the remark this answers is not in the document]
Comment (iwasrobbed) [reply apps/web/pages/export.tsx:241]: Assigning this to @tonyxiao as a one-day edge case in how exports work but not urgent for now

**Review by iwasrobbed — APPROVED** [review]
Good to ship from my end! The rest can get polished along the way 🚀

File: apps/web/pages/export.tsx [hunk 18]
```diff
@@ -111,6 +111,9 @@ interface SyncSpreadsheetCardProps {
 
 function SyncSpreadsheetCard(props: SyncSpreadsheetCardProps) {
   const {apiKey, isEmptyResult, selectedTable} = props
+  if (!selectedTable) {
+    return <></>
```
Comment (bhongy) [inline apps/web/pages/export.tsx:122]: @iwasrobbed fyi `return null` is idiomatic

File: apps/web/pages/api/sql.ts [hunk 19]
```diff
@@ -56,12 +56,13 @@ export default R.identity<NextApiHandler>(async (req, res) => {
       // TODO: Better filename would be nice.
       res.setHeader(
         'Content-Disposition',
-        `attachment; filename="sql-${Date.now()}.${format}"`,
+        `attachment; filename="venice-${Date.now()}.${format}"`,
```
Comment (bhongy) [inline apps/web/pages/api/sql.ts:59]: Nice!

File: apps/web/pages/export.tsx (hunk 18 above)
Comment (iwasrobbed) [reply apps/web/pages/export.tsx]: Great info, thank you for that!
